talk/morelanguage/templates.tex

```diff
@@ -145,27 +146,213 @@
+        template <typename C>
+        void f(typename C::value_type v) { ... }
+        f(std::vector<int>{...}); // cannot deduce C
+                                  // from a dependent type
+        template <typename T, std::size_t N>
+        void g(std::array<T, N * 2> a) { ... }
+        g(std::array<int, 4>{...}); // deduces T=int,
+                                    // cannot deduce N from expression
+        template <typename T>
+        void h(std::vector<T> v) { ... }
+        h({1, 2, 3}); // error, braced-initializer list has no type
+        h(std::vector<int>{1, 2, 3}); // ok, T=int
+      \end{cppcode*}
+    \end{alertblock}
+    \onslide<2>
+    \begin{alertblock}{}
+      \footnotesize
+      \begin{cppcode*}{gobble=2}
+        template <typename C, typename F>
+        void reduce(const C& cont, const F& f = std::plus{});
+        reduce(std::vector<int>{...}); // error: cannot deduce F
+        // would need: <typename C, typename F = decltype(std::plus{})>
+
+        const T & max(const T &a, const T &b) { ... }
+        int i = 3;
+        max(i, 3.14f); // deduces T=int and T=float, error
+        // either: max<float>(i, 3.14f);
+        // or:     max(static_cast<float>(i), 3.14f);
+      \end{cppcode*}
+    \end{alertblock}
+  \end{overprint}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[11]{Template argument deduction}
+  \begin{block}{Deduction in partial specializations}
+    \begin{itemize}
+      \item Partial specializations also deduce template arguments
+      \begin{itemize}
+        \item with similar rules as for function templates
+        \item but some more restrictions (cf.\ \href{https://en.cppreference.com/w/cpp/language/partial_specialization}{cppreference})
+      \end{itemize}
+    \end{itemize}
+  \end{block}
+  \small
+  \begin{cppcode*}{}
+    template <typename T>
+    struct S { ... }; // primary template
+
+    template <typename T>
+    struct S<T*> { ... }; // specialization 1
+
+    template <typename T, std::size_t N>
+    struct S<std::array<T*, N>> { ... }; // specialization 2
+
+    S<int>  s1; // prim. tmpl. (T=int) ok, spec. 1/2 invalid
+    S<int*> s2; // prim. tmpl. (T=int*) and spec. 1 (T=int) ok
+                // spec. 1 is more specialized, will be used
   \end{cppcode*}
 \end{frame}
 
+\begin{frame}[fragile]
+  \frametitlecpp[98]{Template argument deduction}
+  \begin{block}{Specialization vs. overloading}
+    \begin{itemize}
+      \item For function templates we can choose between specialization and overloading
+    \end{itemize}
+  \end{block}
+  \small
+  \begin{multicols}{2}
+    \begin{cppcode*}{}
+      template <typename T>
+      void f(T t) { ... }
+
+
+      void f(int* t) { ... }
+
+
+      template <typename T>
+      void f(T* t) { ... }
+    \end{cppcode*}
+    \columnbreak
+    \begin{cppcode*}{}
+      template <typename T>
+      void f(T t) { ... }
+
+      template <>
+      void f<int*>(int* t) { ... }
+
+      // part. spec. forbidden:
```

Comment:
Should we add why it's forbidden?
We could add this as a bullet point in the box above.

Reply:
So, what is the reason why it is forbidden? :)
The only reasons I know is that it would burden implementers and you can achieve similar effects with overloading.

Reply:
I just added a bullet point that it is forbidden.